core/types/trace.go

```diff
@@ -28,6 +41,41 @@ type TxnTrace struct {
 	StorageReadMap map[libcommon.Hash]struct{}     `json:"-"`
 }
 
+type TxnTraceHex struct {
+	Balance        *Uint256                    `json:"balance,omitempty"`
+	Nonce          *Uint256                    `json:"nonce,omitempty"`
+	StorageRead    []libcommon.Hash            `json:"storage_read,omitempty"`
+	StorageWritten map[libcommon.Hash]*Uint256 `json:"storage_written,omitempty"`
+	CodeUsage      *ContractCodeUsage          `json:"code_usage,omitempty"`
+	SelfDestructed *bool                       `json:"self_destructed,omitempty"`
+	StorageReadMap map[libcommon.Hash]struct{} `json:"-"`
+}
```

Comment:
Why don't we take the same approach as with the `HexBytes`? In the original `TxnTrace` struct, change `*uint256.Int` to a custom struct that you have introduced and then we don't need to replicate the whole `TxnTrace` struct into `TxnTraceHex` (repeat the fields, marshalling logic etc.).

In that case, we would remove the `TxnTraceHex` and just provide the custom marshaling at the level of fields of interest:

```suggestion
```

Reply:
Nice idea! However, I tried that initially before resorting to the current implementation, but it didn't work out. See attached screenshot.
![Screenshot 2024-10-23 at 8 49 25 AM](https://github.com/user-attachments/assets/2ac07ac6-5ad0-4e0d-881a-c5ac1dec1041)
![Screenshot 2024-10-23 at 8 49 03 AM](https://github.com/user-attachments/assets/b4aa9429-19c2-4bfa-9baa-2790609ff8f5)

Reply:
In order for it to work, I had to change all the usage of `uint256` to `Uint256` in https://github.com/0xPolygonHermez/cdk-erigon/blob/zkevm/eth/tracers/native/zero.go, which will incur a bigger scope of code change.

Reply:
Ok, I see...

WDYT about the following (it might need some polishing, but the idea is there). Essentially, the idea is not to fork the entire `TxnTrace` struct, but only those fields that need to be encoded as hex strings:

```go
type Uint256 struct {
    uint256.Int
}

// MarshalJSON implements the json.Marshaler interface
// It returns the hex representation of the uint256.Int.
// We need to do this because the uint256.Int will be marshalled in decimal format by default.
func (u *Uint256) MarshalJSON() ([]byte, error) {
    return json.Marshal(u.Hex())
}

type TxnTrace struct {
    Balance        *uint256.Int                    `json:"balance,omitempty"`
    Nonce          *uint256.Int                    `json:"nonce,omitempty"`
    StorageRead    []libcommon.Hash                `json:"storage_read,omitempty"`
    StorageWritten map[libcommon.Hash]*uint256.Int `json:"storage_written,omitempty"`
    CodeUsage      *ContractCodeUsage              `json:"code_usage,omitempty"`
    SelfDestructed *bool                           `json:"self_destructed,omitempty"`
    StorageReadMap map[libcommon.Hash]struct{}     `json:"-"`
}

// Helper type that only handles custom marshaling for uint256 fields
type TxnTraceHex struct {
    Balance        *Uint256                    `json:"balance,omitempty"`
    Nonce          *Uint256                    `json:"nonce,omitempty"`
    StorageWritten map[libcommon.Hash]*Uint256 `json:"storage_written,omitempty"`
}

// MarshalJSON implements the json.Marshaler interface for TxnTrace.
// It only customizes the marshaling of fields that use uint256.Int.
func (t *TxnTrace) MarshalJSON() ([]byte, error) {
    // Convert the fields that need custom marshaling
    custom := TxnTraceHex{
        StorageWritten: make(map[libcommon.Hash]*Uint256),
    }

    if t.Balance != nil {
        custom.Balance = &Uint256{*t.Balance}
    }
    if t.Nonce != nil {
        custom.Nonce = &Uint256{*t.Nonce}
    }

    for k, v := range t.StorageWritten {
        if v != nil {
            custom.StorageWritten[k] = &Uint256{*v}
        }
    }

    // Use embedding to include the standard fields along with the custom ones
    type alias TxnTrace // Create an alias to avoid infinite recursion with MarshalJSON
    return json.Marshal(struct {
        *alias
        TxnTraceHex
    }{
        alias:         (*alias)(t),  // Marshal the standard fields
        TxnTraceHex: custom,      // Marshal the custom fields
    })
}

```

Reply:
Thanks, it seems an interesting idea! I'd like to, however, keep the current implementation. Having a fully mirrored struct (`TxnTraceHex`) it is a bit redundant conceptually, but it is more readable and understandable to me. The performance should be the same for both implementation, since the rest of fields are simply copied by reference.